opm/simulators/utils/ParallelRestart.cpp

```diff
@@ -0,0 +1,637 @@
+    pack(data.extra, buffer, position, comm);
+}
+
+/// unpack routines
+
+template<class T>
+void unpack(T*, const std::size_t&, std::vector<char>&, int&,
+            Dune::MPIHelper::MPICommunicator, std::integral_constant<bool, false>)
+{
+    OPM_THROW(std::logic_error, "Packing not (yet) supported for this non-pod type.");
+}
+
+template<class T>
+void unpack(T* data, const std::size_t& l, std::vector<char>& buffer, int& position,
+            Dune::MPIHelper::MPICommunicator comm,
+            std::integral_constant<bool, true>)
+{
+#if HAVE_MPI
+    MPI_Unpack(buffer.data(), buffer.size(), &position, data, l,
+               Dune::MPITraits<T>::getType(), comm);
+#endif
+}
+
+template<class T>
+void unpack(T* data, const std::size_t& l, std::vector<char>& buffer, int& position,
+            Dune::MPIHelper::MPICommunicator comm)
+{
+    unpack(data, l, buffer, position, comm, typename std::is_pod<T>::type());
+}
+
+template<class T>
+void unpack(T&, std::vector<char>&, int&,
+            Dune::MPIHelper::MPICommunicator, std::integral_constant<bool, false>)
+{
+    OPM_THROW(std::logic_error, "Packing not (yet) supported for this non-pod type.");
+}
+
+template<class T>
+void unpack(T& data, std::vector<char>& buffer, int& position,
+            Dune::MPIHelper::MPICommunicator comm, std::integral_constant<bool, true>)
+{
+#if HAVE_MPI
+    MPI_Unpack(buffer.data(), buffer.size(), &position, &data, 1,
+               Dune::MPITraits<T>::getType(), comm);
+#endif
+}
+
+template<class T>
+void unpack(T& data, std::vector<char>& buffer, int& position,
+            Dune::MPIHelper::MPICommunicator comm)
+{
+    unpack(data, buffer, position, comm, typename std::is_pod<T>::type());
+}
+
+template<class T1, class T2>
+void unpack(std::pair<T1,T2>& data, std::vector<char>& buffer, int& position,
+            Dune::MPIHelper::MPICommunicator comm)
+{
+    unpack(data.first, buffer, position, comm);
+    unpack(data.second, buffer, position, comm);
+}
+
+template<class T>
+void unpack(std::vector<T>& data, std::vector<char>& buffer, int& position,
+            Dune::MPIHelper::MPICommunicator comm)
+{
+    std::size_t length=0;
+    unpack(length, buffer, position, comm);
+    data.resize(length);
+
+    if (std::is_pod<T>::value)
+    {
+        unpack(data.data(), data.size(), buffer, position, comm);
+        return;
+    }
+
+    for (auto& entry: data)
+        unpack(entry, buffer, position, comm);
+}
+
+void unpack(char* str, std::size_t length, std::vector<char>& buffer, int& position,
+            Dune::MPIHelper::MPICommunicator comm)
+{
+#if HAVE_MPI
+    MPI_Unpack(buffer.data(), buffer.size(), &position, const_cast<char*>(str), length, MPI_CHAR, comm);
+#endif
+}
+
+void unpack(std::string& str, std::vector<char>& buffer, int& position,
+            Dune::MPIHelper::MPICommunicator comm)
+{
+    std::size_t length=0;
+    unpack(length, buffer, position, comm);
+    char* cStr = new char[length];
```

Comment:
just because delete makes me twitch; this could be a vector.

Reply:
I actually agree since I have the tendency to falsely use `delete` instead of `delete[]` for arrays. Will fix.